### DISCUSSION TITLE
Dealing with Messenger queue duplicates

### DISCUSSION THREAD

**Discussion by NiklasBr** [symfony/symfony#49215] [Q&A] [answered]
What would be the best practice to deal with duplicate messages like this?

We have a few `EventListeners`, each of these may trigger the same `Message` to be added to the queue by the `MessageBusInterface $bus`. For example a message to re-index content in the search service after an object has been updated, a user may click save once every few minutes and this may cause many extra rows in the queue table.

Or an API endpoint may update objects irregularly and each update will add the object to the queue for further processing, how do we prevent the same object from being added multiple times?

**Comment by ghost** [comment, marked as answer]
I don't think Symfony Messenger has a built-in solution for this problem. You will have to implement a solution yourself. I would write a middleware that drops the second and every next message for the configured message types. If you run you PHP app in a fire and forget way  (e.g. PHP-FPM) then you can just track the FQCN of the messages in your middleware and drop message if it is a duplicate. However if you run PHP app as long running process (CLI based solutions like PHP-PM or RoadRunner) then you can't just track already sent FQCNs because 2 messages of the same type might be coming from 2 different requests. In this case you will have do some request differentiating. 

If your problem applies also to your cron jobs (console commands) then remember that you don't have request context in this case (but you can still implement some kind of tracking, but in this case not with a use of request). The same with async messenger workers (which are also a console commands) - the message that you don't want to duplicate may be sent multiple times in an async message handler that handles completely different message.

**Reply by NiklasBr** [reply]
> you can still implement some kind of tracking

Do you have any idea where to start with this type of tracking? Is the Header fields in the DB table the right place?

**Reply by ghost** [reply]
I would start with creating a service class that creates and holds a unique id for the request (e.g. inside event listener on kernel.request) and for the currently processing message in case of messenger consumers (inside listener on WorkerMessageReceivedEvent). Then inside a middleware you inject this service and maintain an internal array that groups handled message FQCNs by the unique id provided by this service. 
P.S. Messenger some time ago introduced a batch handling - I don't know if one batch of messages triggers WorkerMessageReceivedEvent for every message inside a batch or just one for the batch. So if you use this feature then please investigate this.

**However** I'm not sure if it works for requests in PHP as long running process (like RoadRunner). Although you receive a new instance of Request object every new request but you may still receive the same instance of TrackingService for two different requests. If so then this service cannot create a new id because another request might not be finished yet and we will be replacing the id it created and is relaying on. But I repeat, I didn't have time to go deep inside such solutions internals and I'm just saying what I read on the Internet (I could misunderstood something).
However if this is the case then I think we can overcome this by setting request id as a custom attribute on Request object and then inside middleware inject a RequestStack service and fetch this attribute's value.
